deep_quoridor/src/agents/alphazero/resnet_network.py

```diff
@@ -0,0 +1,145 @@
+            num_blocks: Number of residual blocks.  Alphazero used 20 for chess and 40 for go, which
+             is a bit more than twice the board dimension. We default to the twice the dimension of
+             the combined grid, plus two.
+            num_channels: Number of channels in convolutional layers (Alphazero used 256)
+        """
+        super(ResnetNetwork, self).__init__()
+
+        self.action_encoder = action_encoder
+        self.device = device
+        self.num_channels = num_channels
+
+        # Calculate input dimensions: MxMx5 where M = board_size * 2 + 3
+        self.input_size = action_encoder.board_size * 2 + 3
+
+        self.num_blocks = num_blocks
+        if self.num_blocks is None:
+            # Alphazero used 20 blocks for chess and 40 for Go, which seems to be a bit more
+            # than double the board dimension.
+            self.num_blocks = self.input_size * 2 + 2
+
+        # Initial convolutional block
+        self.conv_input = nn.Conv2d(5, num_channels, kernel_size=3, padding=1)
+        self.bn_input = nn.BatchNorm2d(num_channels)
+
+        # Residual tower
+        self.residual_blocks = nn.ModuleList([ResidualBlock(num_channels) for _ in range(self.num_blocks)])
+
+        # Policy head
+        self.policy_conv = nn.Conv2d(num_channels, 2, kernel_size=1)
+        self.policy_bn = nn.BatchNorm2d(2)
+        self.policy_fc = nn.Linear(2 * self.input_size * self.input_size, action_encoder.num_actions)
+
+        # Value head
+        self.value_conv = nn.Conv2d(num_channels, 1, kernel_size=1)
+        self.value_bn = nn.BatchNorm2d(1)
+        self.value_fc1 = nn.Linear(self.input_size * self.input_size, 256)
+        self.value_fc2 = nn.Linear(256, 1)
+
+        self.to(self.device)
+
+    def forward(self, x):
+        """
+        Forward pass through the network.
+
+        Args:
+            x: Input tensor of shape (batch_size, 5, M, M) where M = board_size * 2 + 3
+               or tuple of such tensors
+
+        Returns:
+            policy: Policy logits of shape (batch_size, num_actions)
+            value: Value estimate of shape (batch_size, 1) in range [-1, 1]
+        """
+        if isinstance(x, tuple):
+            x = torch.stack([i for i in x]).to(self.device)
+
+        # Initial convolution
+        out = torch.relu(self.bn_input(self.conv_input(x)))
+
+        # Residual tower
+        for block in self.residual_blocks:
+            out = block(out)
+
+        # Policy head
+        policy = torch.relu(self.policy_bn(self.policy_conv(out)))
+        policy = policy.view(policy.size(0), -1)
+        policy = self.policy_fc(policy)
+
+        # Value head
+        value = torch.relu(self.value_bn(self.value_conv(out)))
+        value = value.view(value.size(0), -1)
+        value = torch.relu(self.value_fc1(value))
+        value = torch.tanh(self.value_fc2(value))
+
+        return policy, value
+
+    def game_to_input_array(self, game: Quoridor) -> np.ndarray:
```

Comment:
The combined grid includes an extra 2 layers of solid walls around the board. I left this in since it seems like it could be useful to the convolution kernels.